src/core/thread/mle.cpp

```diff
@@ -4310,100 +4294,198 @@ Mle::DelayedSender::DelayedSender(Instance &aInstance)
+            nextSendTime.UpdateIfEarlier(header.mSendTime);
         }
         else
         {
-            mQueue.Dequeue(message);
-            Send(static_cast<TxMessage &>(message), metadata);
+            Execute(schedule, header);
+            mSchedules.DequeueAndFree(schedule);
         }
     }
 
     mTimer.FireAt(nextSendTime);
 }
 
-void Mle::DelayedSender::Send(TxMessage &aMessage, const Metadata &aMetadata)
+void Mle::DelayedSender::Execute(const Schedule &aSchedule, const Header &aHeader)
 {
-    Error error = kErrorNone;
-
-    aMetadata.RemoveFrom(aMessage);
+    switch (aHeader.mMessageType)
+    {
+    case kTypeDataRequest:
+        IgnoreError(Get<Mle>().SendDataRequest(aHeader.mDestination));
+        break;
 
-    if (aMessage.IsMleCommand(kCommandDataRequest))
+#if OPENTHREAD_FTD
+    case kTypeParentResponse:
     {
-        SuccessOrExit(error = aMessage.AppendActiveAndPendingTimestampTlvs());
-    }
+        ParentResponseInfo info;
 
-    SuccessOrExit(error = aMessage.SendTo(aMetadata.mDestination));
+        IgnoreError(aSchedule.Read(sizeof(Header), info));
+        Get<MleRouter>().SendParentResponse(info);
+        break;
+    }
 
-    Log(kMessageSend, kTypeGenericDelayed, aMetadata.mDestination);
+    case kTypeDataResponse:
+        Get<MleRouter>().SendMulticastDataResponse();
+        break;
 
-    if (!Get<Mle>().IsRxOnWhenIdle())
+    case kTypeLinkAccept:
     {
-        // Start fast poll mode, assuming enqueued msg is MLE Data Request.
-        // Note: Finer-grade check may be required when deciding whether or
-        // not to enter fast poll mode for other type of delayed message.
+        LinkAcceptInfo info;
 
-        Get<DataPollSender>().SendFastPolls(DataPollSender::kDefaultFastPolls);
+        IgnoreError(aSchedule.Read(sizeof(Header), info));
+        IgnoreError(Get<MleRouter>().SendLinkAccept(info));
+        break;
     }
 
-exit:
-    if (error != kErrorNone)
+    case kTypeDiscoveryResponse:
     {
-        aMessage.Free();
+        DiscoveryResponseInfo info;
+
+        IgnoreError(aSchedule.Read(sizeof(Header), info));
+        IgnoreError(Get<MleRouter>().SendDiscoveryResponse(aHeader.mDestination, info));
+        break;
     }
-}
+#endif // OPENTHREAD_FTD
 
-void Mle::DelayedSender::RemoveDataResponseMessage(void)
-{
-    RemoveMessage(kCommandDataResponse, kTypeDataResponse, nullptr);
+    default:
+        break;
+    }
+
+    OT_UNUSED_VARIABLE(aSchedule);
 }
 
-void Mle::DelayedSender::RemoveDataRequestMessage(const Ip6::Address &aDestination)
+bool Mle::DelayedSender::Match(const Schedule &aSchedule, MessageType aMessageType, const Ip6::Address &aDestination)
 {
-    RemoveMessage(kCommandDataRequest, kTypeDataRequest, &aDestination);
+    bool   match = false;
+    Header header;
+
+    header.ReadFrom(aSchedule);
+    VerifyOrExit((header.mMessageType == aMessageType) && (header.mDestination == aDestination));
+    match = true;
+
+exit:
+    return match;
```

Comment:
```suggestion
    Header header;

    header.ReadFrom(aSchedule);
    return (header.mMessageType == aMessageType) && (header.mDestination == aDestination);
```
This could be simpler?

Reply:
Good point! Will change it.